From cf01a43c5eacf6edce607332390041dca900ac6e Mon Sep 17 00:00:00 2001
From: Saurabh Sengar <saurabh.singh@xilinx.com>
Date: Mon, 19 Feb 2018 11:00:26 +0530
Subject: [PATCH 0285/1851] drm: xlnx: mixer: logo layer fix

commit 4ed574455c5881a348776500028412daee713b1b from
https://github.com/Xilinx/linux-xlnx.git

This patch fixes the occasional kernel oops when programing logo layer.
Issue was because of accessing/programing wrong memory.
Following fixes corrected this calculation :
- using the currect buffer b_buf instead of r_buf
- correcting the register address calculation by one offset
- corrected pixel_alpha selection based on color format logic

Signed-off-by: Saurabh Sengar <saurabh.singh@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/xlnx_mixer.c | 13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)

diff --git a/drivers/gpu/drm/xlnx/xlnx_mixer.c b/drivers/gpu/drm/xlnx/xlnx_mixer.c
index 0ca6142711a5..b614d1650e73 100644
--- a/drivers/gpu/drm/xlnx/xlnx_mixer.c
+++ b/drivers/gpu/drm/xlnx/xlnx_mixer.c
@@ -1368,15 +1368,16 @@ static int xlnx_mix_logo_load(struct xlnx_mix_hw *mixer, u32 logo_w, u32 logo_h,
 		shift = (x % 4) * 8;
 		rword |= r_buf[x] << shift;
 		gword |= g_buf[x] << shift;
-		bword |= r_buf[x] << shift;
+		bword |= b_buf[x] << shift;
 		if (mixer->logo_pixel_alpha_enabled)
 			aword |= a_buf[x] << shift;
+
 		if (x % 4 == 3) {
-			reg_writel(reg, (rbase_addr + x), rword);
-			reg_writel(reg, (gbase_addr + x), gword);
-			reg_writel(reg, (bbase_addr + x), bword);
+			reg_writel(reg, (rbase_addr + (x - 3)), rword);
+			reg_writel(reg, (gbase_addr + (x - 3)), gword);
+			reg_writel(reg, (bbase_addr + (x - 3)), bword);
 			if (mixer->logo_pixel_alpha_enabled)
-				reg_writel(reg, (abase_addr + x), aword);
+				reg_writel(reg, (abase_addr + (x - 3)), aword);
 		}
 	}
 
@@ -1423,7 +1424,7 @@ static int xlnx_mix_update_logo_img(struct xlnx_mix_plane *plane,
 		return -EINVAL;
 	}
 	per_pixel_alpha = (logo_layer->hw_config.vid_fmt ==
-			   DRM_FORMAT_RGBA8888) ? false : true;
+			   DRM_FORMAT_RGBA8888) ? true : false;
 	r_data = kcalloc(pixel_cnt, el_size, GFP_KERNEL);
 	g_data = kcalloc(pixel_cnt, el_size, GFP_KERNEL);
 	b_data = kcalloc(pixel_cnt, el_size, GFP_KERNEL);
-- 
2.31.1

